### ISSUE TITLE
[RFC] [docs] Make the availability of version-specific docs more obvious

### ISSUE THREAD

**Issue by jameslamb** [microsoft/LightGBM#5342]
## Summary

LightGBM's official documentation site, https://lightgbm.readthedocs.io/en/latest/, displays documentation built from the latest commit on `master`.

Since the work in #4236, it is also possible to switch to a view pinned to a specific release.

<img width="1436" alt="Screen Shot 2022-06-30 at 10 24 49 PM" src="https://user-images.githubusercontent.com/7608904/176817468-f54a3be1-53d8-4ca9-b0a1-09ed39cf5cec.png">

Changes should be made to make it more obvious which version of `lightgbm` you are viewing the documentation for.

## Motivation

Since it has been 6+ months since the project's most recent release, the documentation built from `master` contains significant deviations from the software that many users have installed (e.g., `pip install lightgbm` returns v3.3.2).

As a result, we have received reports like "the documentation doesn't match the code", e.g. #5113, #5185, #5337.

## Description

I see a few options to make versioning more obvious to users of LightGBM's documentation.

### Option 1 - Change default branch to latest release

As suggested by @jmoralez in https://github.com/microsoft/LightGBM/issues/5337#issuecomment-1170126344, we could edit the settings in readthedocs such that navigating to https://lightgbm.readthedocs.io defaults to the latest release.

<img width="851" alt="image" src="https://user-images.githubusercontent.com/7608904/176818300-e310433a-1066-499f-bdd6-c79bd1187560.png">

This could also be accomplished by pointing to the `stable` git tag, which we move with every release. That is what XGBoost does at https://xgboost.readthedocs.io/.

<img width="1112" alt="Screen Shot 2022-06-30 at 10 49 44 PM" src="https://user-images.githubusercontent.com/7608904/176819859-c50b0660-44cf-43ea-9ed2-29a4aea9f74b.png">

*(the red box is my addition)*

### Option 2 - Include a banner or other indicator warning that you are viewing the `latest` docs

For example, using something like https://github.com/humitos/sphinx-version-warning (although that specific project was last updated in 2019).

Looks like this:

<img width="1108" alt="image" src="https://user-images.githubusercontent.com/7608904/176819630-aa5d8ed9-98e7-4a73-bdd4-c260aa7b8b63.png">

ref: https://sphinx-version-warning.readthedocs.io/en/latest/

### Option 3 - Choose a different Sphinx theme where the version is more prominent visually

See examples at https://sphinx-themes.org/#themes

## Notes for Reviewers

@StrikerRUS @jmoralez @guolinke @shiyu1994 Whenever you have time, please add an opinion.

**Comment by jameslamb** [comment]
I personally support Option 1, switching the default version that https://lightgbm.readthedocs.io/ directs to to the `stable` tag.

**Comment by jmoralez** [comment]
I'm +1 on the stable tag

**Comment by guolinke** [comment]
I think we can make the default one to stable as well

**Comment by StrikerRUS** [comment]
Building docs for `stable` branch fails with the following warning:
```
Running Sphinx v5.0.2
WARNING: Invalid configuration value found: 'language = None'. Update your configuration to a valid langauge code. Falling back to 'en' (English).
...
Updating searchtools for Read the Docs search... build finished with problems, 1 warning.
```

Also, we have to refactor the following code for R-package.
https://github.com/microsoft/LightGBM/blob/a52d9c3097404325790ef08777700563faa2f5c0/docs/_static/js/script.js#L11-L22

I believe the easiest way to workaround this is to set default branch at RTD site manually after each new release.

**Comment by jameslamb** [comment]
Ah I see, I didn't consider the fact that we don't currently build `stable` at readthedocs, so it's not guaranteed to have been successfully built before.

Ok yeah! I think manually updating the default branch after each release is a good solution. We check the docs on release branches prior to merging release PRs and adding a tag, so we'll know the docs have built successfully.

Since you are an admin on RTDd and I'm not ... would you update RTD to point to the v3.3.2 docs by default?

**Comment by StrikerRUS** [comment]
Or maybe we can start building `stable` tag at RTD with `4.0.0` to avoid the need to backport some Sphinx hotfixes.

**Comment by StrikerRUS** [comment]
> would you update RTD to point to the v3.3.2 docs by default?

Done.

![image](https://user-images.githubusercontent.com/25141164/178157730-4eeac467-5881-4baf-bec5-62998cafb5e0.png)

**Comment by jameslamb** [comment]
Very nice, thanks @StrikerRUS ! I can see now when I go to https://lightgbm.readthedocs.io/ that it redirects to v3.3.2.

> Or maybe we can start building stable tag at RTD with 4.0.0 to avoid the need to backport some Sphinx hotfixes.

Oh good point! I agree, that would help. And then if we could rely on `stable` being the default, we wouldn't have to do anything manual to move the version docs are pinned to when we do a new release.

I added a note about that to the description of #5153.

I think now that we've pinned to a specific version and won't possibly change to `stable` until v4.0.0, this specific issue can be closed.

Thanks to everyone for the help!
